articles/200_coexistence_and_migration.md

```diff
@@ -0,0 +1,247 @@
+
+#### Build type
+
+Almost all build types for ROS 1 and ROS 2 are disjoint except the build type `cmake` which can be processed by both build tools.
+Since it is not desired to change existing `catkin` / `ament_cmake` packages to use the build type `cmake` those are currently different for each ROS version.
+
+### CMake version
+
+ROS 2 requires CMake version 3.5+.
+It should be easily possible to update existing CMake code to be compatible with various versions of CMake.
+
+### Python version
+
+ROS 2 requires Python version 3.5+.
+It should be easily possible to update existing Python code to be compatible with Python 2.7 as well as Python 3.5+.
+
+### CMake API
+
+ROS 1 packages use the CMake API provided by [catkin](https://github.com/ros/catkin).
+ROS 2 packages use a slightly different CMake API provided by [ament_cmake](https://github.com/ament/ament_cmake).
+The rational for the changes and the different name are described in [this article](http://design.ros2.org/articles/ament.html).
+
+**Idea:** Create a ROS 2 package names `catkin` which provides the same CMake API as the ROS 1 package (minus unavailable features like the devel space) by mapping the calls to `ament_cmake` API.
+In addition it is expected that existing packages need to call a new function at the end of their CMake code to account for the fact that `ament_package` needs to be invoked *after* all targets where `catkin_package` needs to be invoked *before* them.
+
+### Message / service definitions
+
+In ROS 2 message / service definitions are expected to be in the subdirectory `msg` / `srv`.
+This implicitly ensures that message and service names are unique.
+There is no technical reason behind that design decision rather than simplifying the implementation.
+
+**Idea:** Allow arbitrary locations in the source space.
+
+Currently ROS 2 already supports additional feature in these files (e.g. default values) which are not available in ROS 1.
+In the future ROS 2 might support a different file format to satisfy the requested feature requests (like grouping constants in enums, optional fields, file / constant / field comments, etc.).
+If that is the case a backward compatible parser to read existing `.msg` / `.srv` files will likely be available.
+
+### Generated message / service API
+
+In order to satisfy the goal of dual-homed applications the generated code of ROS 1 and ROS 2 need to be side-by-side usable.
+
+For C++ that is already the case.
+In ROS 1 the header files are generated in namespace specific to the package name.
+In ROS 2 the header files are placed into a sub-namespace names `msg` / `srv`.
+Beside the different path / namespace the data structure also uses different types for e.g. the shared pointers (Boost vs. C++11).
+
+For Python that is not the case.
+
+**Proposal:** Change the location of the generated Python modules from using `msg` / `srv` as the subdirectory to `message` / `service`.
```

Comment:
This sounds fine to me.

I remember there were some comments a while back to the effect of changing the include paths increases the migration effort. It may be possible to create a ROS2 package that uses `importlib` to import only from ROS 1 packages. This would mean import statements would be unchanged when porting from ros1 to ros2, but dual homed applications could use something extra to get ROS 1 python packages.

Dual homed code might look something like

```
from my_pkg.msg import MyMsg as MyMsgRos2
MyMsgRos1 = ros1_bridge.import('my_pkg.msg', 'MyMsg')
```

Reply:
I also remember seeing that people are bothered by the change in C++ header namespaces, and we explained that it was a requirement for the dual-homed applications. However, in [that conversation](https://github.com/ros2/design/pull/50) I didn't see an explanation of why it would be technically necessary in python. If it's true that it could be worked around in python (perhaps by what @sloretz has suggested, or something else), it seems that this proposal conflicts with the 'minimise migration effort' goal, and the namespaces should be left as is in python

Reply:
> Dual homed code might look something like

```
from my_pkg.msg import MyMsg as MyMsgRos2
MyMsgRos1 = ros1_bridge.import('my_pkg.msg', 'MyMsg')
```

@sloretz This snippet would not work since the imported file identified by the arguments `my_pkg.msg` and `MyMsg` might also call import and would therefore resolve to the ROS 2 version of the dependent modules (assuming the message generation hasn't been altered).

Reply:
@dirk-thomas How about registering/unregistering an [import hook](https://www.python.org/dev/peps/pep-0302/) using a `with` statement?

```
from my_pkg.msg import MyMsg as MyMsgRos2
with ros1_bridge.Ros1Import:
    from my_pkg.msg import MyMsg as MyMsgRos1
```

Reply:
@sloretz An import hook might help with this but I haven't used it. Please consider providing an example which demonstrates loading of a ROS 1 as well as ROS 2 message in the same Python script, e.g. `geometry_msgs/PointStamped` to cover the case of fields being of other message types.